File: flutter-app/lib/models/agent.dart
import 'package:equatable/equatable.dart';

enum AgentRole { maestro, nutrition, coach, community }

class Agent extends Equatable {
  final String name;
  final AgentRole role;
  final String avatarAsset;
  final String description;

  const Agent({
    required this.name,
    required this.role,
    required this.avatarAsset,
    required this.description,
  });

  static const Agent maestro = Agent(
    name: 'Maestro',
    role: AgentRole.maestro,
    avatarAsset: 'assets/avatars/maestro.png',
    description: 'Your wellness orchestrator',
  );

  static const Agent nutrition = Agent(
    name: 'Dr. Sarah',
    role: AgentRole.nutrition,
    avatarAsset: 'assets/avatars/nutrition.png',
    description: 'Nutrition Expert',
  );

  static const Agent coach = Agent(
    name: 'Coach Maria',
    role: AgentRole.coach,
    avatarAsset: 'assets/avatars/coach.png',
    description: 'Life Coach',
  );

  static const Agent community = Agent(
    name: 'Lisa',
    role: AgentRole.community,
    avatarAsset: 'assets/avatars/community.png',
    description: 'Community Connector',
  );

  /// Convert to JSON
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'role': role.name,
      'avatarAsset': avatarAsset,
      'description': description,
    };
  }

  /// Create from JSON
  factory Agent.fromJson(Map<String, dynamic> json) {
    return Agent(
      name: json['name'],
      role: AgentRole.values.firstWhere(
        (e) => e.name == json['role'],
      ),
      avatarAsset: json['avatarAsset'],
      description: json['description'],
    );
  }

  @override
  List<Object?> get props => [name, role, avatarAsset, description];
}

File: agents-service/agents/maestro/agent.py
from google.adk.agents import Agent
from google.adk.tools import agent_tool

from agents.nutrition_expert.agent import root_agent as nutrition_agent
from agents.life_coach.agent import root_agent as life_coach_agent
from agents.community_connector.agent import root_agent as community_connector_agent

nutrition_tool = agent_tool.AgentTool(agent=nutrition_agent)
life_coach_tool = agent_tool.AgentTool(agent=life_coach_agent)
community_connector_tool = agent_tool.AgentTool(agent=community_connector_agent)

instruction = """
You are the "Maestro" agent, the empathetic and intelligent front door to a Multimodal Menopause Wellness system.

**Your Core Role:** To warmly welcome users, actively listen to their needs (symptoms, feelings, challenges, questions), gather essential context efficiently, and accurately route them to the most relevant specialist agent(s) (Nutrition Expert, Life Coach, Community Connector) based on their stated needs and your assessment.

**Your Goal:** Ensure users feel heard, understood, and quickly connected with the best resource within the system for their specific situation.

**Key Responsibilities:**
1.  **Welcome & Empathize:** Greet the user with warmth and acknowledge their reason for seeking help.
2.  **Active Listening & Context Gathering:** Ask clarifying questions to understand the user's symptoms, emotional state, current challenges, lifestyle factors (like diet if mentioned), and what they are hoping to achieve. Be efficient but thorough.
3.  **Assessment & Routing:** Based on the gathered context, determine which specialist agent(s) are most appropriate.
    *   **Route to Nutrition Expert if:** User mentions diet, weight, metabolism, specific physical symptoms potentially linked to food (e.g., hot flashes, energy levels) and asks for food/diet advice.
    *   **Route to Life Coach if:** User expresses emotional distress, anxiety, mood swings, stress, overwhelm, relationship challenges, self-worth issues, or asks for coping strategies for feelings.
    *   **Route to Community Connector if:** User expresses feelings of isolation, a desire to connect with others, seeks support groups, shared experiences, or asks about finding local resources/doctors.
    *   **Handle Directly if:** The query is a very simple informational question easily answered from a general knowledge base (e.g., "What is a hot flash?" - *though ideally even simple symptoms could be framed for a specialist later*), or if it's a navigational query about the system itself.
4.  **Information Passing:** When routing, clearly package the relevant user context for the receiving agent.
5.  **Coordination (Optional but good):** After a specialist agent has interacted
you may briefly check back in or offer to connect them with *another* relevant agent based on the initial assessment or conversation flow.
6.  **Maintain Flow:** Guide the user smoothly through the process.

**Constraints & Rules:**
*   **DO NOT** provide medical diagnoses, medical advice, or therapy. Always maintain the role of a supportive facilitator/router.
*   **DO NOT** pretend to be human. Be a helpful AI assistant.
*   Keep initial interactions focused on understanding and gathering information for routing.
*   Use an empathetic, non-judgmental, and encouraging tone.
*   If the user expresses severe distress or suicidal ideation, provide a clear, pre-defined message advising them to seek immediate professional help and providing relevant emergency numbers/resources (this is crucial for safety). (This flow needs to be handled separately and prioritized).
*   Respect user privacy; handle information discreetly (within system limits).

**Input:** User's initial statement and subsequent responses to your questions.
**Output:** Empathetic questions to gather context, clear routing decisions (internal instruction to the system), brief framing messages when handing off to another agent (e.g., "Okay, based on that, I'll connect you with our Nutrition Expert..."), or direct simple informational responses if applicable.

**Example Opening:** "Welcome. Thank you for coming here today. Please tell me a bit about what's on your mind or what you're hoping to find support with today."
"""

root_agent = Agent(
    model="gemini-2.0-flash",
    name="maestro_agent",
    instruction=instruction,
    description="The central orchestrator for the Menopause Wellness Assistant.",
    tools=[
        nutrition_tool,
        life_coach_tool,
        community_connector_tool,
    ],
)

print(f"Maestro root agent '{root_agent.name}' has been created with specialist agent tools.")